### ISSUE TITLE
Reorganize plotting functions

### ISSUE THREAD

**Issue by dcdanko** [ewels/MultiQC#182]
To continue discussion about new plotting functionality in #180 and #179 and reorganization in #181:

Could you (we?) consider breaking out plotting functionality a bit? To me it seems like it would be easier to understand the different types of plots available if each type of plot was implemented as its own class.

In my opinion this would make it easier to:

- Understand the available types of plots
- Integrate modules with plots
- Create new types of plots
- declutter the base module files

BUT I learned to code in Java so I do tend to over use classes.

**Comment by ewels** [comment]
Yes, I think this is a good idea - the base module file is getting kind of huge. I'm not particularly fussed about making them into classes (I learned to code in Perl so I tend to under use classes!) but it would definitely be good to separate this code into separate files at least.

I'm hoping to have a session on some core code updates at the end of this week - beeswarm plot, heatmap, tables and this code organisation all fits together rather nicely. I'll see if I can wrap it all up into one set of changes.

How about having a multiqc python submodule called plots? Then we could create plots within modules as follows:

```
multiqc
├── modules
├── templates
├── plots
│   ├── __init__.py
│   ├── bargraph.py
│   ├── beeswarm.py
│   ├── linegraph.py
│   ├── heatmap.py
│   └── table.py
└── utils
```

```python
from multiqc.plots import bargraph, linegraph

html += bargraph.plot(data, cats)
html += linegraph.plot(data)
```

Sound sensible?

Phil

**Comment by dcdanko** [comment]
That makes perfect sense to me. That's essentially how I implemented arbitrary tables in #180 albeit that the class was under multiqc/utils/

**Comment by ewels** [comment]
Great - I have meetings all day tomorrow but I'll try to finish off the beeswarm plots and then do this reorganisation after that. Then I'll get started on adding heatmaps and tables. Once that's done, I'll refactor the general stats code to use this centralised table code. Sounds like a plan!

Thanks for the suggestion..

Phil
